server/src/main/java/org/elasticsearch/health/node/DiskHealthIndicatorService.java

```diff
@@ -157,8 +157,9 @@ static class DiskHealthAnalyzer {
             for (String nodeId : diskHealthByNode.keySet()) {
                 DiscoveryNode node = clusterState.getNodes().get(nodeId);
                 HealthStatus healthStatus = diskHealthByNode.get(nodeId).healthStatus();
-                // TODO #90213 update this only after we check that this health status indicates a problem.
-                if (mostSevereStatusSoFar.value() < healthStatus.value()) {
+                // we don't want the disk indicator status to change to UNKNOWN solely because one node couldn't report
+                // its disk health status
+                if (healthStatus.indicatesHealthProblem() && mostSevereStatusSoFar.value() < healthStatus.value()) {
```

Comment:
What do you think if we moved this below `if (node == null || healthStatus.indicatesHealthProblem() == false)`. This way we would only count the statuses that indicate a problem but also we would only count the nodes that exist in the cluster state.

Reply:
I'd prefer to have it explicit ie. the definition of the most severe status has changed with this PR (ie. the health status needs to indicate a problem - and this is now expressed in where we compute the most severe status) as opposed to implicit ie. we checked that `healthStatus.indicatesHealthProblem() == false` as part of a different check above 

> but also we would only consider the status of the nodes that exist in the cluster state

Hm, this is interesting. I think we _should_ consider the node's status irrespective of its presence in the cluster state (the disk indicator will report it right?)

Reply:
Ah thinking about it some more, this is the output we can get with the current implementation where the status is computed based on nodes in the cache, but the `symptom` is based on the nodes in the cluster state (also the counts under `details` are based on the cache values irrespective of the cluster state).
```
  "disk": {
      "status": "red",
      "symptom": "0 nodes with roles: [] are out of disk or running low on disk space.",
      "details": {
        "blocked_indices": 0,
        "green_nodes": 1,
        "unknown_nodes": 0,
        "yellow_nodes": 0,
        "red_nodes": 1
      }
    },
```

The validation against the local cluster state comes from the need to create a categorisation based on roles. 
The health API coordinating node is expected to have a reasonably up to date cluster state (due to the `master_is_stable` indicator being a pre-flight indicator) so the case when a node is in the cache but not in the cluster state can arise if:
1) the health API coordinating node hasn't heard _yet_ of the new cluster membership, but the health node received a health status from a new node
2) the health API coordinating node has the latest cluster membership but the coordinating node already received the health cache overview using the previous membership configuration

So moving forward, coming back to the fact that we trust the coordinating node's cluster state to be up-to-date enough to serve health information data from it directly, I think your suggestion makes sense - ie. don't let the nodes that are not in the cluster state affect the worse seen status and we need to follow-up and fix the `details` section to not account for the nodes that are not in the cluster state.

What do you think?

Reply:
Opened https://github.com/elastic/elasticsearch/issues/90715